src/core/components/pin/gc-lock.js

```diff
@@ -0,0 +1,135 @@
+'use strict'
+
+const mortice = require('mortice')
+const pull = require('pull-stream')
+const EventEmitter = require('events')
+const log = require('debug')('ipfs:gc:lock')
+
+class GCLock extends EventEmitter {
+  constructor (repoOwner) {
+    super()
+
+    // Ensure that we get a different mutex for each instance of GCLock
+    // (There should only be one GCLock instance per IPFS instance, but
+    // there may be multiple IPFS instances, eg in unit tests)
+    const randId = (~~(Math.random() * 1e9)).toString(36) + Date.now()
+    this.mutex = mortice(randId, {
+      singleProcess: repoOwner
+    })
+
+    this.lockId = 0
+  }
+
+  readLock (lockedFn, cb) {
+    return this.lock('readLock', lockedFn, cb)
+  }
+
+  writeLock (lockedFn, cb) {
+    return this.lock('writeLock', lockedFn, cb)
+  }
+
+  lock (type, lockedFn, cb) {
+    if (typeof lockedFn !== 'function') {
+      throw new Error(`first argument to ${type} must be a function`)
```

Comment:
```suggestion
      return cb(new Error(`first argument to ${type} must be a function`))
```

Reply:
The reason I throw here instead of returning an error in the callback is that not providing the right arguments to the function is a programmer error. Do you think that's the right approach?

Reply:
Elsewhere we use the [`assert`](https://nodejs.org/api/assert.html) module to guard against programmer errors.

Reply:
Thanks, that's a better idea 👍

Reply:
👍 yep good point